tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -957,112 +957,87 @@ unsigned CGMSHLSLRuntime::ConstructStructAnnotation(
 
   unsigned CBufferSize = CBufferOffset;
 
-  for (RecordDecl::field_iterator Field = RD->field_begin(),
-                                  FieldEnd = RD->field_end();
-       Field != FieldEnd;) {
-    if (Field->isBitField()) {
-      // TODO(?): Consider refactoring, as this branch duplicates much
-      // of the logic of CGRecordLowering::accumulateBitFields().
-      DXASSERT(CGM.getLangOpts().HLSLVersion > hlsl::LangStd::v2015,
-               "We should have already ensured we have no bitfields.");
-      CodeGenTypes &Types = CGM.getTypes();
-      ASTContext &Context = Types.getContext();
-      const ASTRecordLayout &Layout = Context.getASTRecordLayout(RD);
-      const llvm::DataLayout &DataLayout = Types.getDataLayout();
-      RecordDecl::field_iterator End = Field;
-      for (++End; End != FieldEnd && End->isBitField(); ++End)
-        ;
-
-      std::vector<DxilFieldAnnotation> BitFields;
-
-      RecordDecl::field_iterator Run = End;
-      uint64_t StartBitOffset = Layout.getFieldOffset(Field->getFieldIndex());
-      uint64_t Tail = 0;
-      for (; Field != End; ++Field) {
-        uint64_t BitOffset = Layout.getFieldOffset(Field->getFieldIndex());
-        // Zero-width bitfields end runs.
-        if (Field->getBitWidthValue(Context) == 0) {
-          Run = End;
-          continue;
+  // By this point the structType in the StructAnnotation has been stripped of
+  // excess fields if it is a union. As a result we must construct the
+  // annotation differently for unions and structs
+  if (RD->isUnion() && !RD->field_empty()) {
+    FieldDecl *fieldDecl = nullptr;
+    bool largestIsBitfield = false;
+    unsigned largestBitwidth = 0;
+
+    for (FieldDecl *Field : RD->fields()) {
+      CBufferOffset = 0;
+
+      QualType fieldTy = Field->getType();
+
+      CBufferOffset = AlignBaseOffset(fieldTy, CBufferOffset, bDefaultRowMajor,
+                                      CGM, dataLayout);
+
+      const hlsl::ConstantPacking *packOffset = nullptr;
+      for (const hlsl::UnusualAnnotation *it : Field->getUnusualAnnotations()) {
```

Comment:
It doesn't look like any of the tests apply unusual annotations to union members, and it seems to me like you shouldn't be able to specify a pack offset on a union member because all union members should start at the same pack offset.

The (extremely inadequate) spec that I wrote [here](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0004-unions.md) explicitly disallowed semantics on union fields, we should probably also extend that to disallow all HLSL annotations (semantics, pack offset, register, etc).

We should make sure to have parsing tests to verify that we get errors too.